File: 剑指offer/36.二叉搜索树与双向链表.py
class Node:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def treeToDoublyList(self, root: 'Node') -> 'Node':
        def help(root):
            nonlocal p
            if not root:
                return

            help(root.left)
            p.right = root
            root.left = p
            p = root
            help(root.right)

        if not root:
            return None

        new_head = Node(0)
        p = new_head
        help(root)
        new_head.right.left = p
        p.right = new_head.right
        return new_head.right

File: 剑指offer/36.二叉搜索树与双向链表.cpp
#include <iostream>
#include <vector>

using namespace std;

class Node {
public:
    int val;
    Node* left;
    Node* right;

    Node() { }

    Node(int _val)
    {
        val = _val;
        left = NULL;
        right = NULL;
    }

    Node(int _val, Node* _left, Node* _right)
    {
        val = _val;
        left = _left;
        right = _right;
    }
};

class Solution {
public:
    void help(Node* root, vector<Node*>& nodes)
    {
        if (!root) {
            return;
        }

        help(root->left, nodes);
        nodes.push_back(root);
        help(root->right, nodes);
    }

    Node* treeToDoublyList(Node* root)
    {
        vector<Node*> nodes;
        help(root, nodes);

        if (nodes.size() == 0) {
            return NULL;
        }

        Node* head = nodes[0];
        Node* pre = nodes[nodes.size() - 1];

        for (int i = 0; i < nodes.size(); i++) {
            Node* cur = nodes[i];

            Node* next = NULL;
            if (i == nodes.size() - 1) {
                next = head;
            } else {
                next = nodes[i + 1];
            }

            cur->right = next;
            cur->left = pre;

            pre = cur;
        }

        return head;
    }
};